### ISSUE TITLE
pod behavior when calico is down

### ISSUE THREAD

**Issue by niroowns** [projectcalico/calico#6113]
Hi,

We're seeing that egress from a pod start up was broken when the corresponding calico-node pod on the same node within a EKS cluster was "down" (it was up, but not ready due to liveliness probes failing).  I have two questions:

1.  What is the expected behavior for pods that happen to be on calico-nodes that are not operating properly? In this case, the pods had no corresponding network policies associated with them at all.

2.  We're noticing that pods are flapping due to liveliness probe timeouts.  This particular cluster is largish (1300 pods, 60 nodes).  I've seen this issue: https://github.com/projectcalico/calico/issues/5122#issuecomment-982865898.  I am wondering if we need to bump up requests/limits for the calico-node ds?

## Your Environment
* Calico version: node: v3.20.0, tigera operator: v1.20.1
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubernetes (EKS 1.21)
* Operating System and version: Amazon Linux 2

**Comment by caseydavenport** [comment]
> What is the expected behavior for pods that happen to be on calico-nodes that are not operating properly? In this case, the pods had no corresponding network policies associated with them at all.

If calico/node stops running on a particular node, the main things that won't happen are respecting changes to the current network topology. For example, new pods that are launched on that node won't have the necessary rules programmed to enable them to access the network. Similarly, changes to network policy won't be reflected until after Calico starts again. 

A slightly less obvious consequence is that new pods launching on other healthy nodes may not get network access to pods on the unhealthy node, since the bad node is unable to update its policy rules to account for the newly launched remote pods (assuming there is policy in place).

All of that said, the existing pods and policies will continue to function as normal while Calico is down, it's only newly launched pods / changes to policy that should be impacted. 

> This particular cluster is largish (1300 pods, 60 nodes)

In Calico terms, this isn't especially large (we've scale tested up to 5k nodes in the past, and pod densities of 100+ pods per node). However, if you have resource limits on Calico you definitely want to check that they are appropriate for your cluster - it's hard to say how many resources a given cluster will use, because it depends on many many different factors.

**Comment by niroowns** [comment]
Hi @caseydavenport,

Thanks for the reply.  On the statement that new application pods are impacted, is this true even if these pods have 0 network policies associated with them? I am trying to understand if calico-node, at start up, defaults to a blocking posture? Or is this more because this was an existing node and calico was running on it at some point in time and later became unhealthy?

On the liveliness probe failure, in this particular cluster, CPU and memory is really overcommitted on certain nodes, so I think there may be cpu contention going on causing the liveliness probe to flap.

**Comment by caseydavenport** [comment]
> is this true even if these pods have 0 network policies associated with them? 

That's correct, even with no policies Calico defaults to a blocking posture. From Calico's perspective, if something is wrong, we don't know if there are or aren't policies that should be programmed for this pod, and so the only safe thing to do is deny traffic by default.

**Comment by caseydavenport** [comment]
> So just bumping the requests/limits may only be a band aid.

Perhaps, although ensuring all your applications have requests / limits set will help Kubernetes to avoid over-committing your nodes. You may still need to add more resources to your cluster, but it will fail reliably rather than intermittently and unpredictably.

**Comment by niroowns** [comment]
thanks @caseydavenport - this has been helpful.

**Comment by caseydavenport** [comment]
Going to close this for now but feel free to open another issue or respond here if you have any other questions. You can also check out our Slack channel for more interactive discussion
